src/scipp/curve_fit.py

```diff
@@ -0,0 +1,520 @@
+      >>> z = sc.linspace(dim='zz', start=0.0, stop=1, num=10)
+      >>> y = z * 0 + func(x, a=5, b=17/sc.Unit('m'))
+      >>> y.values += 0.01 * rng.normal(size=500).reshape(10, 50)
+      >>> da = sc.DataArray(y, coords={'x': x, 'z': z})
+
+      >>> popt, _ = curve_fit(
+      ...    ['x'], func, da,
+      ...    p0 = {'b': 1.0 / sc.Unit('m')}, reduce_dims=['zz'])
+      >>> round(sc.values(popt['a']), 3), round(sc.stddevs(popt['a']), 4)
+      (<scipp.DataArray>
+       Dimensions: Sizes[]
+       Data:
+                             float64  [dimensionless]  ()  5
+       ,
+       <scipp.DataArray>
+       Dimensions: Sizes[]
+       Data:
+                             float64  [dimensionless]  ()  0.0021
+       )
+
+      Note that the variance is about 10x lower in this example than in the
+      first 1D example. That is because in this example 50x10 points are used
+      in the fit while in the first example only 50 points were used in the fit.
+
+    """
+
+    if 'jac' in kwargs:
+        raise NotImplementedError(
+            "The 'jac' argument is not yet supported. "
+            "See https://github.com/scipp/scipp/issues/2544"
+        )
+
+    for arg in ['xdata', 'ydata', 'sigma']:
+        if arg in kwargs:
+            raise TypeError(
+                f"Invalid argument '{arg}', already defined by the input data array."
+            )
+
+    for c in coords:
+        if c in da.coords and da.coords.is_edges(c):
+            raise BinEdgeError("Cannot fit data array with bin-edge coordinate.")
+
+    if not isinstance(coords, dict):
+        if not all(isinstance(c, str) for c in coords):
+            raise TypeError(
+                'Expected sequence of coords to only contain values of type `str`.'
+            )
+        coords = {c: c for c in coords}
+
+    # Mapping from function argument names to fit variables
+    coords = {
+        arg: da.coords[coord] if isinstance(coord, str) else coord
+        for arg, coord in coords.items()
+    }
+
+    p0 = _make_defaults(f, coords.keys(), p0)
+    p_units = [p.unit if isinstance(p, Variable) else default_unit for p in p0.values()]
+
+    f = (
+        _wrap_scipp_func(f, p0, p_units)
+        if not unsafe_numpy_f
+        else _wrap_numpy_func(f, p0, coords.keys())
+    )
+
+    map_over = tuple(
+        d
+        for d in da.dims
+        if d not in reduce_dims and not any(d in c.dims for c in coords.values())
+    )
+
+    dims_participating_in_fit = set(da.dims) - set(map_over)
+
+    # Create a dataarray with only the participating coords and masks
+    # and coordinate names matching the argument names of f.
+    _da = DataArray(
+        da.data,
+        coords=coords,
+        masks={
+            m: da.masks[m]
+            for m in da.masks
+            if dims_participating_in_fit.intersection(da.masks[m].dims)
+        },
+    )
+
+    for m in _da.masks.values():
+        _da = _da[~m]
```

Comment:
Won't this fail if a mask is not 1D? I don't think there is an easy way to fix this, just wondering if you thought about this, or added a test?

Reply:
Hm, I did not think about that.
One alternative is to only apply the 1d masks and then set values masked by 2D masks to `nan`. The Scipy `curve_fit` function will then ignore the missing values.

Reply:
I would not change anything until we have a concrete requirement, maybe just add a test to ensure that it fails as expected and does not silently do something wrong.

Reply:
Fixed it by applying the masks to the flattened dataarray instead of to the nd-array.